WIP: Extend spline documentation

danielwolff1 opened the pull request:
# Overview
Started to adapt the spline documentation. 

## Addressed issues
*  #132 

## Checklists
* [ ] Documentations are up-to-date.
* [ ] Added example(s)
* [ ] Added test(s)

danielwolff1 commented:
This is WIP for now but since I did not get feedback if the amount of documentation is too detailed or too superficial, I wanted to prepare this merge request so that you have easier comparison, I'll keep working on that.

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,57 @@
     """Bezier Base. Contain extra operations that's only
     available for bezier families.
+
+    Bezier splines are a special type of spline where the basis functions have
```
```suggestion
    Beziers are a special type of spline where the basis functions have
```

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,57 @@
+            N_{i;p}(u) = \\binom{p}{i} u^i (1-u)^{p-i}
+
+    A Bezier spline itself can be seen as a map from the parametric domain
+    :math:`\\[0,1\\]^{N_{param}}` to the physical domain
```
```suggestion
    :math:`[0,1]^{N_{param}}` to the physical domain
```
Original did not work on my machine, please confirm that this is still valid on mac.

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,57 @@
+    of splines depending on the dimension of the parametric space.
+
+    #. A spline of degree :math:`p` with control points
+    :math:`P_i\\in\\mathbb{R}^{N_{phys}` and a one-dimensional parameter space
```
```suggestion
    :math:`P_i\\in\\mathbb{R}^{N_{phys}}` and a one-dimensional parameter space
```

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,57 @@
+            C(u) = \\sum_{i=0}^{l} N_{i;p}(u) P_i
+
+    #. A spline of degrees :math:`p,q` with control points
+    :math:`P_{i,j}\\in\\mathbb{R}^{N_{phys}` and a two-dimensional parameter
```
```suggestion
    :math:`P_{i,j}\\in\\mathbb{R}^{N_{phys}}` and a two-dimensional parameter
```

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,57 @@
+    often rewritten as in terms of multi-variate basis functions
+
+    .. math::
+            \\tilde{N}_{i,j;p,q}(u,v) \\coloneqq N_{i;p}(u) N_{j;q}(v)
```
```suggestion
            \\tilde{N}_{i,j;p,q}(u,v) := N_{i;p}(u) N_{j;q}(v)
```
Coloneqq is unknown to mathjax, please use := instead

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,57 @@
+
+    #. A spline of degrees :math:`p,q,r` with control points
+    :math:`P_{i,j,k}\\in\\mathbb{R}^{N_{phys}` and a three-dimensional
+    parameter space corresponds to a surface, embedded into the physical space:
```
```suggestion
    parameter space corresponds to a volume, embedded into the physical space:
```
I think we got the idea with surfaces already, but this can stay. Please also mention, that parametric and physical domain can be different in a general case and that we support up to 10 dimensions for both the parametric and physical domain seperately

danielwolff1 replied:
In the documentation, they are not assumed to be equal, that's why different symbols for the parametric and physical domains are used. But I'm now going to add additional information to the documentation and store them in the `Spline` class.

jzwar commented:
![image](https://github.com/tataratat/splinepy/assets/63095162/1c4c4259-d9ca-4d72-a252-4989e9ee291e)
For reference

danielwolff1 commented:
Thanks for the fast review! 

Since I am now finally able to build the documentation locally, I will fix the issues with the rendering asap!

danielwolff1 commented on splinepy/bezier.py:
```diff
@@ -171,6 +222,9 @@ def __init__(self, degrees=None, control_points=None, spline=None):
         """
         Bezier (Spline).
 
+        See :class:`.BezierBase` for more information on the theory of Bezier
+        splines.
+
         Parameters
         -----------
         degrees: (para_dim,) list-like
```
<img width="755" alt="Screenshot 2023-05-24 at 10 09 14" src="https://github.com/tataratat/splinepy/assets/39346676/607dc8dd-24a8-47dd-a76d-f91fb68c869f">

Are you sure? For me it displays perfectly fine, or I don't get what you mean... 🤔

jzwar replied:
There are two blocks of docstring. The one attached to the constructor is not displayed. That's why I think the docstring is supposed to be written into the class header

jzwar replied:
Check out the Bezier class fir reference

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,61 @@
+    *Berstein polynomials*
+
+    .. math::
+            B_{i;p}(u) = \\binom{p}{i} u^i (1-u)^{p-i}
```
You can do without the double backslashes by using raw strings. 
```python 
r"""
Something with \
"""

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,61 @@
 class BezierBase(spline.Spline):
     """Bezier Base. Contain extra operations that's only
     available for bezier families.
+
```
Can you please add the "s" in the line above, I know it was probably my mistake, but I just noticed it :wink:

j042 replied:
Probably my mistake. Also "'re"  (`Contains extra operations that're --`)

danielwolff1 replied:
Sorry guys, I didn't spot the mistakes while extending the documentation, otherwise I would have changed it for sure, now I'm gonna adapt it after this review :)

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,61 @@
+    We can distinguish between different types of splines depending on the
+    dimension of the parametric space.
+
+    1. A spline of degree :math:`p` with control points
```
```suggestion
    1. A spline of degree :math:`p` with :math:`(p+1)` control points
```

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,61 @@
+    physical space:
+
+    .. math::
+            S^B(u) = \\sum_{i=0}^{l} B_{i;p}(u) P_i
```
This is just a minor change, but I suggest to stick to the general naming convention used by GE and also NURBS book, where curves are `C`, Surfaces `S` and Volumes `V`, higher dimensional objects are then written as `M` for multivariate.

I don't mind if you stick to the current notation, just wanted to put it out there.

danielwolff1 replied:
Alright, I can change that.

jzwar commented on splinepy/bezier.py:
```diff
@@ -165,6 +220,13 @@ def composition_derivative(self, inner, inner_derivative):
+
+    See :class:`.BezierBase` for more information on the theory of Bezier
+    splines.
+    """
```
Please also move the init docstring here so the parameters are described in the documentation.

j042 replied:
sphinx default is not to document any dunder methods. I once added a workaround to document `__init__` and it seems that I removed it in 646741a. Would you rather just revive that or document all of them in class doctring?

jzwar replied:
I don't think an explicit documentation of the `__init__` and other python specifics makes a lot of sense. I'd rather have the documentation within the class header.

danielwolff1 replied:
Okay, then I'll move the `__init__()` methods to the class doc-strings to resolve this.

jzwar commented on splinepy/bezier.py:
```diff
@@ -4,6 +4,61 @@
+    Finally, we like to emphasize that Bezier splines can also be seen as a
+    special type of B-Splines with open knot vectors (i.e., the first and last
+    entry are repeated :math:`p+1`-times) that do not feature additional
+    internal knots.
```
What do you think of adding a simple example to show how it is to be used:
```suggestion
    internal knots.
    
    Usage:
    
    .. code-block:: python
       
       # Polynomial Bezier surface
       polynomial_bezier = splinepy.Bezier(
           degrees=[2,1],
           control_points=[
               [0.0, 0.0],
               [1.0, 0.0],
               [2.0, 1.0],
               [0.0, 2.0],
               [1.0, 1.0],
               [2.0, 2.0]
           ]
       )
       # Rational bezier surface (quarter arc)
       rational_bezier = splinepy.RationalBezier(
           degrees=[2],
           control_points=[
               [0.0, 0.0],
               [1.0, 0.0],
               [1.0, 1.0]
           ],
           weights=[1.0, 2.0**-0.5, 1.0]
       )
```

danielwolff1 replied:
Sure, I can do that, I'll copy examples from the examples folder.

jzwar commented on splinepy/bspline.py:
```diff
@@ -4,8 +4,82 @@
+    physical space:
+
+    .. math::
+            S^N(u,v) = \sum_{i=1}^{l} \sum_{j=1}^{m} N_{i;p}(u) N_{j;q}(v)
```
See above

jzwar commented on splinepy/nurbs.py:
```diff
@@ -2,8 +2,83 @@
+    .. math::
+            W^N(u) = \sum_{i=1}^{l} N_{i;p}(u) w_i
+
+    with scalar weights :math:`w_i\in\mathbb{R}^{> 0}`.
```
```suggestion
    with scalar weights :math:`w_i\in\mathbb{R}^{> 0}`. Note, that the same basis functions are used for both the weighting function and projection space.
```

jzwar commented on splinepy/bspline.py:
```diff
@@ -4,8 +4,82 @@
+
+    Here, :math:`l` denotes the number of basis functions/control points along
+    this parametric dimension and :math:`p` the polynomial order of these
+    basis functions.
```
I know it's a bit of an overkill, but can you add the relation of the expected number of control points, knots and the degree respectively

jzwar reviewed (commented):
Thank you so much for taking the time to extend the documentation. Just a couple of general remarks:
1. The notation used here differs in some aspects from the one used in mapper, especially when it comes to the upper index notation. I don't really care which notation is used (semi-colon or upper index) so long as it is consistent.
2. Some of the links are missing and it might be good to have a short commentary on what dimensions are supported, just in case someone (ehm - like me) tries to use 64-dimensional Splines.
3. I personally like the idea of including small examples of how the different splines are generated, just so you don't have to jump back and forth to find the appropriate example. This can be a separate PR, in which case please write an issue.

Once these changes are implemented, this can be approved.

danielwolff1 commented on splinepy/bspline.py:
```diff
@@ -4,8 +4,82 @@
 
 class BSplineBase(spline.Spline):
-    """BSpline base. Contains extra operations that's only available for
+    r"""BSpline base. Contains extra operations that are only available for
```
Shouldn't that be clear from the fact that `NURBS` inherits from `BSplineBase`? But sure, I can add that comment

danielwolff1 commented on splinepy/nurbs.py:
```diff
@@ -2,8 +2,83 @@
+    NURBS are an extension of B-Splines overcoming their drawback of being
+    unable to represent circular shapes. This is achieved by the introduction
+    of a weighting function
+    :math:`W^N:\mathbb{R}^{N_{param}}\to\mathbb{R}^{> 0}`, defined as (for
```
What index supposed to mean, I haven't seen that before :D

danielwolff1 commented:
Thanks for the extensive review guys. I'll address your points and come back to you.

danielwolff1 commented:
I added examples and addressed (most) points of the review. Next, I'm gonna adopt the notation to the one in `mapper`, i.e., basically, I'm gonna replace all subscripts with superscripts. This is going to look weird, but it's the easiest non-intrusive fix to make the notations compatible. 

(For the theory part it is important to understand which basis function has which degree, while this is not required when writing down the equations for transforming derivatives between the parametric and the physical domain. If I only made the summation index a superscript and let the polynomial degree as a subscript, it would not be consistent because the mapper class uses subscripts only for components).

j042 reviewed (approved):
Again, this is awesome. Thank you so much @danielwolff1! I will just add minor format fix and merge
